### ISSUE TITLE
RFC: Standartize content prop

### ISSUE THREAD

**Issue by layershifter** [Semantic-Org/Semantic-UI-React#1364]
I'm think that we need to remove `content` prop from all our components.
This will be breaking change.

### Confusing

It's represents different types in our components:
- [contentShorthand](https://github.com/Semantic-Org/Semantic-UI-React/blob/060e81db9c18e8cb8d20b804abafa75e1e8579e8/src/views/Feed/FeedContent.js#L66)
- [itemShorthand](https://github.com/Semantic-Org/Semantic-UI-React/blob/060e81db9c18e8cb8d20b804abafa75e1e8579e8/src/views/Feed/FeedEvent.js#L65)

### Poor coverage

There are many components that doesn't have `content` prop. [Sometime](https://github.com/Semantic-Org/Semantic-UI-React/blob/8c619da2dc2e94e0e1c4f09746e6c42ed1adeab4/src/elements/Step/StepDescription.js#L42) they have another shorthand instead of `content`.

### Useless

It does same job as `children` prop:

```jsx
<Component>Children</Component>
<Component children='Children' />
<Component content='Children' />
```

### Poor tests

We doesn't have special tests for `contentShorthand`, our [existing](https://github.com/Semantic-Org/Semantic-UI-React/blob/060e81db9c18e8cb8d20b804abafa75e1e8579e8/test/specs/views/Feed/FeedContent-test.js#L40) is not so powerfull as `rendersChildren`.

### Simplify render function

Our `render` funtions will be more simple, while we have less usages of `lodash` (`_.isNil`).

```diff
- {_.isNil(children) ? content : children}
+ {children}
```

-------------

@levithomason @jcarbo I'm glad to hear your throughts :smile

**Comment by levithomason** [comment]
I agree on confusing, poor coverage, and simplifying the render, however, it is not only useful but required to facilitate the shorthand factories.

The purpose of the `content` prop is that it supports shorthand creation.  This is different than `children` in two very important ways:

1. Children does not support props objects
1. Children does not create an element, but renders the value as-is

Example, a Popup takes `content` which ultimately creates a `content` div within the Popup:

```jsx
<Popup content='Some content' />

<div class="ui popup transition visible">
  <div class="content">Some content</div>
</div>
```

Whereas using `children` would, of course, render the text directly as a child.  We also cannot pass `children` to the factory in every case because there would be no way for the user to pass certain values, such as an array of strings or a props object.

There are some usages which are unnecessary, such as the ItemMeta's content.  This usage is a direct replacement of the `children` prop and offers no new features.  I think the only reason we should keep these props is for API consistency and this design principal:

_All components should have a `content` prop and it should create the required primary content of the component from a number, string, props object, or another element_.

I think then we must keep the content prop for shorthand purposes.  Let me know your thoughts on this.

**Comment by layershifter** [comment]
> There are some usages which are unnecessary, such as the ItemMeta's content. This usage is a direct replacement of the children prop and offers no new features.

I want to eliminate possible misunderstandings. My proposal is remove `content: contentShorthand` that makes same job as `children` (`ItemMeta`), while keep `content: itemShorthand` (`Popup`) that is part of our design.

**Comment by levithomason** [comment]
I can see that argument.  Keeping consistent props across all components is nice, however, our shorthand props are 1:1 with subcomponents.  So, if there is no `Foo.Content` subcomponent, then it is reasonable to expect there should be no `<Foo content />` prop.

I was leaning toward removing the prop for this case, however, I have one other consideration.  Shorthand props, such as `icon` are incompatible with `children` for longstanding [reasons](https://github.com/Semantic-Org/Semantic-UI-React/pull/335#issuecomment-238960895).  If a component has shorthand props and the user wants to use them in conjunction with `children`, then there is no way to do so:

```jsx
<Button icon='user'>Profile</Button>          // does not work

<Button icon='user' content='Profile' />      // works
```

Essentially, the `content` prop is shorthand for `children` if you are using shorthand props.

We could allow the use of `children` in place of the `content` prop in most of these cases.  My only concern there is the cases where we cannot do this so easily.  Cases where the `content` is used with logic to change the markup and/or used as a value to `createShorthand`.  See [Message](https://github.com/Semantic-Org/Semantic-UI-React/blob/master/src/collections/Message/Message.js#L184) which passes `content` to create a shorthand paragraph, this would not work with passing children.  Also, [Header](https://github.com/Semantic-Org/Semantic-UI-React/blob/master/src/elements/Header/Header.js#L78)'s shorthand content which has a Content subcomponent but `content` is used to alter the markup and not for shorthand Content creation, we don't want to alter the markup every time there are children.  What if the user wants exactly the children that they passed?

The inconsistency issue is also raised again, that as a developer I have no way of knowing if a component can use `children` or must use `content` with my shorthand props.  This goes back to the longstanding [reasons conclusion](https://github.com/Semantic-Org/Semantic-UI-React/pull/335#issuecomment-238960895).

***

In conclusion, I'm thinking it is still good to keep `content` for all components for consistency and predictability:

- `children` are always rendered exactly as you passed them, never altered/wrapped
- shorthand props are never compatible with children
- every component has a `content` prop that is the primary content
- all components behave the same all the time

**Comment by layershifter** [comment]
Main problem that I see there is incompability :confused: 

```jsx
<Popup content='Some content' />

<div class="ui popup transition visible">
  <div class="content">Some content</div>
</div>

<Popup content={{ children: 'Some content' }} />

<div class="ui popup transition visible">
  <div class="content">Some content</div>
</div>
```

```jsx
<Button icon='user' content='Profile' />

<button class="ui button">
  <i class="user icon"></i> Profile
</button>

<Button icon='user' content={{ children: 'Profile' }} /> // does not work
```

**Comment by levithomason** [comment]
Oh boy, this is a great point.  We simply cannot have some content props work with shorthand, and others not.  Exploring ideas here, it is possible to create a component from this:

```jsx
<Button content={{ children: 'Profile' }} />

// => <button class="ui button"><span>Profile</span></button>
```

This would mean that all usages of the content prop _must_ be passed to `createShorthand`.  We are bound to run into CSS issues since `content='foo'` would also create an element.

I'll have to think on this one.  At this point, it seems the least destructive path is to keep the `content` prop but also allow inconsistent shorthand usage.  I don't like this idea, let's think on it and look for something better.

**Comment by layershifter** [comment]
I have some proposals.

### content and contentRenderer

We have `renderer` functions in some components. It's a quite good pattern, I think we can use it there:

```jsx
function PopupContent(props) {
  // ...
 return <ElementType>{_.isNil(children) ? contentRender(content) : children}</ElementType>
}

PopupContent.defaultProps = {
  contentRenderer: defaultContentRenderer
}

PopupContent.propTypes = {
  // ...
  content: customPropTypes.contentShorthand,
  contentRenderer: Proptypes.func,
}
```

```jsx
const defaultContentRenderer = content => {
  if(typeof content === 'string' || typeof content === 'number') return content
  return createShorthand('div', content)
}
```

So, if we have `string` or `number` on `content` we will simply return them. If we have `object`, let's use shorthand power, but we will wrap it into child element.

### rename content prop

Yes, it will be breaking, but `content` on `Popup` and `PopupContent` still will make different things. It's a good  step to clarify usage of shorthands.

**Comment by levithomason** [comment]
>So, if we have `string` or `number` on content we will simply return them. If we have object, let's use shorthand power, but we will wrap it into child element.

I like this on first glance except for two hang ups:

1. Added complexity
2. There is no way to know which element to wrap it with.  A `div` cannot be a child of may components.  A `span` might be safer, but it still may break CSS in some cases.

I'm now thinking that we do not worry about this until we have reported issues about it.  I personally have never run into it.  Do you have any use cases or bugs you've run into that requires a refactor?
